cmd/combine.go

```diff
@@ -0,0 +1,62 @@
+// Copyright © 2022 Obol Labs Inc.
+//
+// This program is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the Free
+// Software Foundation, either version 3 of the License, or (at your option)
+// any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of  MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package cmd
+
+import (
+	"context"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+
+	"github.com/obolnetwork/charon/combine"
+)
+
+func newCombineCmd(runFunc func(ctx context.Context, lockfile, inputDir, outputDir string) error) *cobra.Command {
+	var (
+		inputDir  string
+		outputDir string
+		lockfile  string
+	)
+
+	cmd := &cobra.Command{
+		Use:   "combine",
+		Short: "Combines private key shares into a single private key for a distributed validator.",
+		Long:  "Combines private key shares into a single private key for a distributed validator.\nWarning: running the resulting private key in a validator alongside the original distributed validator will result in slashing.",
+		Args:  cobra.NoArgs,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return runFunc(cmd.Context(), lockfile, inputDir, outputDir)
+		},
+	}
+
+	bindCombineFlags(
+		cmd.Flags(),
+		&inputDir,
+		&outputDir,
+		&lockfile,
+	)
+
+	return cmd
+}
+
+func newCombineFunc(ctx context.Context, lockfile, inputDir, outputDir string) error {
+	return combine.Combine(ctx, lockfile, inputDir, outputDir)
+}
+
+func bindCombineFlags(flags *pflag.FlagSet, inputDir, outputDir, lockfile *string) {
+	flags.StringVar(inputDir, "keyfile-dir", "./validator-keys", `Directory containing all the "keyfile-N.json" and "keyfile-N.txt" files`)
+	flags.StringVar(outputDir, "out-dir", "./recombined-validator-key", "Directory where to save the resulting private key")
+	flags.StringVar(lockfile, "lockfile-path", "./lock.json", "Charon lock.json file")
```

Comment:
suggest aligning with `run` flag: `lock-file`

Reply:
How does this work for e.g. a 10 validator, 4 node cluster? Do you put 40 files in a flat folder? or 4 subfolders? do we do any traversing/exploring subdirectories?

Reply:
I believe this command should be require some work from the user: to recombine all the keys, it's assumed some work is needed to arrange them in the correct order.

If we want to automate some part of this, we could (as you suggested) let the user provide the `validator_keys` directories renamed with the public key share, and do some directory traversing to recombine.

I still believe though that each validator recombination needs a separate invocation of the `combine` command.

Reply:
Mmmm similar to exits, I believe in practice we will see people mostly combine/exit all validators in a cluster or none. I think partial will be a rare edge case. With the scale of enterprise deployments, I don't think a separate invocation per key is scalable, everyone will DIY their own shell script for-loop. 

I think @Battenfield should user test some options, but I think 'people will send their `.charon` folder to the combiner, the combiner will dump all `n` folders into one directory, having to rename each of them something other than .charon (awkward but feels unavoidable), and from there I think the should be able to run `charon combine --data-dir ./this-folder-ive-prepared`, from where we should be able to traverse it, assuming that each subdir is a .charon folder, and we can find `./this-folder-ive-prepared/*/cluster-lock.json` in each of them, and even a `./this-folder-ive-prepared/*/validator_keys/*` path that we can use to find hopefully all of the keystores we expect. Assuming we can find everything we need, get or create `output-dir` and spit out the combined keystores.

Reply:
> everyone will DIY their own shell script for-loop

Correct, I implied users will do something similar... I don't see it as being completely unfeasible, and on the flipside depending on deployments needs it could actually be more comfortable.

I imagine a K8S deployment where validator keys are secrets, you have to script your way around the secrets API to mount directories in pods and have `combine` work with them anyway.

Regardless, I'm open to changes on this.

Reply:
Yeah, you make me also wonder should we add a `keymanager-address` flag like the DKG and create cluster command, that can push the combined keys into a web3 keymanager. 

I think humour me on this one and we go with this `combine` command attempting to combine every private key in the lockfile and warns/(errors and needs to be flag overridden) if it can't discover a threshold of the private keys and passwords it needs to complete.

Reply:
@corverroos @OisinKyne considering this process yields standard private keys, I think it's better to write a file with the private key bytes to disk instead of our own keystore files: if a group of operator wants to recombine their private key(s), most probably they want to stop use charon, so writing keystore files seems a bit pointless.

Thoughts?

Reply:
@gsora the intention behind writing a keystore is to align with an ethereum standard here https://eips.ethereum.org/EIPS/eip-2335 This is a format supported by all  clients. The idea is they would import these keystores into the clients as if they were exported from the eth2-staking-cli. Our `.charon/validator_keys` folder also matches that structure (albeit with key *shares* not 'normal' keys)

Reply:
Ok, I’ll write keystore files then. 

The rest of the UX is fine for you?